fix(site): fix react state violation in filetree create/update utils

DanielleMaywood opened the pull request:
Closes https://github.com/coder/coder/issues/20482

Ensure file tree utils don't modify their parameters so we don't violate Reacts rules

DanielleMaywood committed:
fix(site): react state violation in filetree create/update utils

Closes https://github.com/coder/coder/issues/20482

I've just replaced `set` from `lodash` with `set` from `lodash/fp`. We
could alternatively just ensure `fileTree` is cloned before using the
mutable version.

BrunoQuaresma reviewed (commented):
That is interesting. Could we create a test for that? To make sure we don't have any regressions in the future?

buenos-nachos reviewed (commented):
The change itself is right – I'm just worried about what the style of change means for the larger codebase

DanielleMaywood committed:
chore: respond to feedback

- Replace use of `lodash/fp/set` with `lodash/cloneDeep` followed by
`lodash/set`.
- Fix `removeFile` as that still did mutation.
- Update tests to assert that the original FileTree was not modified.

DanielleMaywood committed:
chore: appease the formatter

DanielleMaywood committed:
fix: use structuredClone instead of cloneDeep

DanielleMaywood commented on site/src/utils/filetree.ts:
```diff
@@ -1,3 +1,4 @@
+import cloneDeep from "lodash/cloneDeep";
```
Addressed in https://github.com/coder/coder/pull/20483/commits/416b55c78f5b48f679fe8d11a408b08c0d0da15a
